packages/core-data/src/types/index.ts

```diff
@@ -0,0 +1,105 @@
+/**
+ * Internal dependencies
+ */
+import { Attachment } from './attachment';
+import { Comment } from './comment';
+import { MenuLocation } from './menu-location';
+import { NavMenu } from './nav_menu';
+import { NavMenuItem } from './nav_menu_item';
+import { NavigationArea } from './navigation-area';
+import { Page } from './page';
+import { Plugin } from './plugin';
+import { Post } from './post';
+import { Settings } from './settings';
+import { Sidebar } from './sidebar';
+import { Taxonomy } from './taxonomy';
+import { Theme } from './theme';
+import { User } from './user';
+import { Type } from './type';
+import { Widget } from './widget';
+import { WidgetType } from './widget-type';
+import { WpTemplate } from './wp_template';
+import { WpTemplatePart } from './wp_template_part';
+import { Context, RenderedText } from './common';
+
+export {
+	Attachment,
+	Comment,
+	MenuLocation,
+	NavMenu,
+	NavMenuItem,
+	NavigationArea,
+	Page,
+	Plugin,
+	Post,
+	Settings,
+	Sidebar,
+	Taxonomy,
+	Theme,
+	User,
+	Type,
+	Widget,
+	WidgetType,
+	WpTemplate,
+	WpTemplatePart,
+};
```

Comment:
This is nice, but exporting only the `type` versions of these (all the ones that have been passed through `OmitNevers`, which should be all of them in my opinion, see related comment for that) will disallow plugin authors from extending them.

There are two solutions I could see for this:

1. Export the `Full`-prefixed interfaces by themselves, call it a day.
2. Change the "full" version of these types to be inside a namespace (`global.wpCoreData`) so that they can be extended through the global namespace.

Re: suggesting using namespaces... some TS fans might be rolling their eyes (or worse, maybe they're hexing me!) but I think in this specific case it's an appropriate way to expose these "base" types that would need to be extended by plugin authors without adding them to the "public API" of the package. For example, it's nice for them not to show up when you do this:

```ts
import type { Attachment } from '@wordpress/core-data`';
```

It'd be nice if there's only a single "canonical" `Attachment` type that you can import directly from the library. This reduces confusion about which one to use. If `FullAttachment` and `Attachment` were both present, many consumers of the package would be tempted by the power that `FullAttachment` seems to imply :slightly_smiling_face:

Reply:
I'm not rolling my eyes, but I'm concerned about breaking tooling and builds and introducing problems others will have to deal with. I'm not knowledgable here so I'm just rehashing things I've read and asking based off of my limited understanding. Noted in a comment above.

Reply:
I'll mark this one as resolved and let's keep discussing in the open thread above.